### ISSUE TITLE
API Server package does per-request mTLS

### ISSUE THREAD

**Issue by coltonstapper** [kubernetes/kubernetes#90533]
**What would you like to be added**:
The API Server's secure serving settings today [only requests the client cert](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go#L64), and its [x509 Authenticator and Verifier implementations verify the cert signing chain of every request](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go#L116), even after mTLS has already taken place upon connection handshake.

As an improvement, if the secure serving, Authenticator, and Verifier implementations were changed so that the API Server's TLS config verified the cert upon initial connection (VerifyClientCertIfGiven instead of RequestClientCert), and the auth libraries used the VerifiedChains of the negotiated connection, then all requests streamed through that connection would not be doing mTLS anymore.

To preserve backwards compatibility we could implement a new Authenticator/Verifier implementation that's feature gated in the generic API Server options.

**Why is this needed**:
TL;DR this will save an enormous amount of CPU that's being wasted on extra auth for every request.

Per request auth isn’t necessary as it’s not going to change during the establishment of the connection. If we are worried about the certificate expiring, then we can timebox the length of the connection (as most keepalive TLS connections do) and cause the client to reconnect forcing a re-authorization. If we are worried about long term polling style connections potentially expiring their certs (or being revoked), the best solution is again, timebox and cause them to reconnect. 

We can do all sorts of verification if we really want to on each request (though, I would argue, just like other systems, once you’ve authorized yourself to them, you have that entitlement for whatever period you’re connected) without doing the full EC or RSA math required to validate the signatures of the certificates. That math won’t change.

**Additional Info**:
This mainly bit us because our clusters were using 384bit ECDSA root certs, which is currently not decrypted in constant time in Go's standard crypto library: https://github.com/golang/go/issues/11499

The effect was that API Servers effectively run out of CPU once scaled to several hundred thousand requests per minute, because of the per-request mTLS auth logic.

Once switched to per-connection auth, the issue went away. Here's a graph showing CPU usage on the API Servers before and after the patch:

<img width="1708" alt="Screen Shot 2020-03-18 at 10 10 04 AM" src="https://user-images.githubusercontent.com/8000471/80413203-f1614b00-8883-11ea-907c-f21d71959307.png">

Even though it seems other Kubernetes users aren't experiencing reliability issues with the extra auth in their clusters because they're using cheaper certs, I'd imagine that many organizations using Kubernetes would want the reduced CPU from this relatively simple change. 

I haven't run scale tests using different certs, but I'd be interested if anyone has also internally switched to only per-connection mTLS and seen the performance improvements!

I'm creating the issue to kick off discussion around if I'm missing something here as to why this is the connection behavior today, and also to propose some solutions. Feedback is super welcome, thanks!

**Comment by coltonstapper** [comment]
This is a simple example of the patch we used, though this breaks other cluster's use cases because it assumes that there are no intermediary certs/different CA certs so some logic has to change there to support those. But I think shows the general idea of what the config options and the Authenticator/Verifier would look like, stripping the verify() calls and using the VerifiedChains from the already verified connection:
```
diff --git a/cmd/kubelet/app/server.go b/cmd/kubelet/app/server.go
index 8a705bb139a..d63da0234fe 100644
--- a/cmd/kubelet/app/server.go
+++ b/cmd/kubelet/app/server.go
@@ -1008,7 +1008,7 @@ func InitializeTLS(kf *options.KubeletFlags, kc *kubeletconfiginternal.KubeletCo
 		// Specify allowed CAs for client certificates
 		tlsOptions.Config.ClientCAs = clientCAs
 		// Populate PeerCertificates in requests, but don't reject connections without verified certificates
-		tlsOptions.Config.ClientAuth = tls.RequestClientCert
+		tlsOptions.Config.ClientAuth = tls.VerifyClientCertIfGiven
 	}
 
 	return tlsOptions, nil
diff --git a/staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go b/staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go
index da3ef45d26c..7aeb62653d7 100644
--- a/staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go
+++ b/staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go
@@ -112,28 +112,11 @@ func (a *Authenticator) AuthenticateRequest(req *http.Request) (*authenticator.R
 		return nil, false, nil
 	}
 
-	// Use intermediates, if provided
-	optsCopy, ok := a.verifyOptionsFn()
-	// if there are intentionally no verify options, then we cannot authenticate this request
-	if !ok {
-		return nil, false, nil
-	}
-	if optsCopy.Intermediates == nil && len(req.TLS.PeerCertificates) > 1 {
-		optsCopy.Intermediates = x509.NewCertPool()
-		for _, intermediate := range req.TLS.PeerCertificates[1:] {
-			optsCopy.Intermediates.AddCert(intermediate)
-		}
-	}
-
 	remaining := req.TLS.PeerCertificates[0].NotAfter.Sub(time.Now())
 	clientCertificateExpirationHistogram.Observe(remaining.Seconds())
-	chains, err := req.TLS.PeerCertificates[0].Verify(optsCopy)
-	if err != nil {
-		return nil, false, err
-	}
 
 	var errlist []error
-	for _, chain := range chains {
+	for _, chain := range req.TLS.VerifiedChains {
 		user, ok, err := a.user.User(chain)
 		if err != nil {
 			errlist = append(errlist, err)
@@ -174,22 +157,6 @@ func (a *Verifier) AuthenticateRequest(req *http.Request) (*authenticator.Respon
 		return nil, false, nil
 	}
 
-	// Use intermediates, if provided
-	optsCopy, ok := a.verifyOptionsFn()
-	// if there are intentionally no verify options, then we cannot authenticate this request
-	if !ok {
-		return nil, false, nil
-	}
-	if optsCopy.Intermediates == nil && len(req.TLS.PeerCertificates) > 1 {
-		optsCopy.Intermediates = x509.NewCertPool()
-		for _, intermediate := range req.TLS.PeerCertificates[1:] {
-			optsCopy.Intermediates.AddCert(intermediate)
-		}
-	}
-
-	if _, err := req.TLS.PeerCertificates[0].Verify(optsCopy); err != nil {
-		return nil, false, err
-	}
 	if err := a.verifySubject(req.TLS.PeerCertificates[0].Subject); err != nil {
 		return nil, false, err
 	}
diff --git a/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go b/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go
index 6d165abdf55..37b86167e76 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go
@@ -61,7 +61,8 @@ func (s *SecureServingInfo) tlsConfig(stopCh <-chan struct{}) (*tls.Config, erro
 	if s.ClientCA != nil {
 		// Populate PeerCertificates in requests, but don't reject connections without certificates
 		// This allows certificates to be validated by authenticators, while still allowing other auth types
-		tlsConfig.ClientAuth = tls.RequestClientCert
+		// Always verify the cert upon handshake if given
+		tlsConfig.ClientAuth = tls.VerifyClientCertIfGiven
 	}
 
 	if s.ClientCA != nil || s.Cert != nil || len(s.SNICerts) > 0 {
```

**Comment by liggitt** [comment]
Optimizing to do the certificate verification once per connection could make sense (especially if it is bounded to the lifetime of the cert), but switching to have the http.Server do the verification via `VerifyClientCertIfGiven` would break clients which present an unrecognized client certificate at connection time and then present valid bearer tokens in requests.

**Comment by liggitt** [comment]
> Per request auth isn’t necessary as it’s not going to change during the establishment of the connection.

Note that some API server authentication configuration (including [CA](https://github.com/kubernetes/kubernetes/pull/83579) [bundles](https://github.com/kubernetes/kubernetes/pull/85004)) can be dynamic, so it is possible for the results of an x509 validation to change within the lifetime of a connection.

**Comment by coltonstapper** [comment]
@liggitt hmm, for the first point, why would we want to support clients that are presenting unrecognized client certs if they're trying to do bearer token auth? If the client's trying to obtain a token it'd present no cert (I believe this is what kube-controller-manager does, for example).

For the second point, if the CA bundle changes we would need to disconnect any connections that authorized previously so they can reauthorize.

**Comment by enj** [comment]
> why would we want to support clients that are presenting unrecognized client certs if they're trying to do bearer token auth?

Well, we always have at this point, so it may be too late to ask _why_.  Generally speaking, we try all ways to authenticate a request in a specific order, and pick the first one that works.  So short-circuiting because client cert verification fails is not the desired behavior.

**Comment by liggitt** [comment]
> why would we want to support clients that are presenting unrecognized client certs if they're trying to do bearer token auth? If the client's trying to obtain a token it'd present no cert

That is not universally true. We've observed real clients configured to present client certs unconditionally at a transport level (e.g. browsers or clients expecting to speak to corporate proxies that require client certs) in addition to whatever kubernetes-specific bearer tokens are required to authenticate to the API server.

**Comment by liggitt** [comment]
Ideally, golang would support a VerifyClientCertIfGiven-style option that would continue with empty VerifiedChains on cert verification failure.

A few years ago, I explored moving the verify from request-level to connection-level while preserving the ability to serve requests presenting unknown client certificates, and wasn't able to find a reasonable way to do it without reimplementing significant portions of the TLS/http stack. It's possible some of the additions to go make this doable now, so proposals or suggestions are welcome.

**Comment by liggitt** [comment]
I actually wonder if a golang option would be the most reasonable approach here, long-term. PeerCertificates+VerifiedChains in the request would mean verification succeeded. PeerCertificates without VerifiedChains would mean it failed.

**Comment by coltonstapper** [comment]
@liggitt understood, yeah I completely understand preserving the existing use case to not break clients. It makes sense that a Go http change would allow us to implement the desired change while maintaining existing behavior for clients doing different types of auth, though that does seem like it'll take longer to upstream.

Are you opposed to the idea of a separate request Authentication implementation being a flag as a short term solution while I kick off the Golang discussion? I think this is on the SIG Auth agenda for Wednesday so we could talk about it there too.

**Comment by liggitt** [comment]
I'd prefer to keep the authentication implementation in one place and avoid adding user-facing options for this... the test matrix for the apiserver is already extremely large and I'm not confident these new options would have good coverage.

A possible short-term mitigation could be to add a LRU verification cache to the authenticator that remembers the last N certificate validation results. The cache key could be a hash of the presented peercertificates and the generation of the validation options (static validation options would have a generation of 1, dynamic validation options could bump their generation when a changed CA bundle was observed, etc). That assumes computing something like sha1 on the peercertificates would be significantly faster than than crypto verification (would need benchmarking and proving out).

**Comment by enj** [comment]
I agree with Jordan on this.  I am happy to improve performance here, but there must be no changes to user facing config.

The options I see are:

1. Carry a patch in your own fork of the API server code (not a great long term solution)
1. Avoid P-384 (not realistic)
1. Fix Go's P-384 implementation (hard because crypto)
1. Implement `AttemptToVerifyClientCertIfGiven` (I do not think it would not be hard to implement, but the bar for changes to Go's stb lib are about as high as it gets and the Go team may have reservations for such a change)
1. Add some caching to the API server's cert verification (doable w/o config knobs but would require benchmarking and the code may be complex)

**Comment by mikedanese** [comment]
> 5\. Add some caching to the API server's cert verification (doable w/o config knobs but would require benchmarking and the code may be complex)

This is my preference

> LRU verification cache to the authenticator

Please no more LRUs :). Do we care about invalidation if we are re-verifying certs e.g. every 10 seconds or 1 minute?

**Comment by coltonstapper** [comment]
Discussing in the sync today, it's understood that adding a feature flag doesn't exactly make sense if there's no plan to converge the code to the main implementation (which we can't do because it breaks clients), and the Golang route may take a long time, so the cache approach is the best short to medium term one!

> Please no more LRUs :). Do we care about invalidation if we are re-verifying certs e.g. every 10 seconds or 1 minute?

@mikedanese we'll cache the verified hash and the hash dates, which means we can set the cache expiration for the entry as the cert expiration time to kill two birds with one stone, if that sounds good? We can set a configured max size for the cache so that users can adjust depending on how much memory they want to dedicate to it (or 0 to turn off the cache altogether?).

Also, thinking about it, in a follow up PR we could support a negative cache as well, so that a particular client can't DDOS the API anymore because the failed hash would get rejected right away, instead of having to re-verify at each connection time :) this would solve the problem we had with an important misconfigured client that was taking our API Server down with a bunch of invalid requests.

**Comment by coltonstapper** [comment]
Discussing in the sync today, it's understood that adding a feature flag doesn't exactly make sense if there's no plan to converge the code to the main implementation (which we can't do because it breaks clients), and the Golang route may take a long time, so I agree the cache approach is the best short to medium term one!

**Comment by mikedanese** [comment]
Checkout https://github.com/kubernetes/apimachinery/blob/e0747e0da69d35c4c63520fd646fd2f839168bb6/pkg/util/cache/expiring.go#L42 potentially with striping like we do in cached_token_authenticator.go.

**Comment by enj** [comment]
This cert cache needs to support audit annotations, i.e. #90140.

**Comment by coltonstapper** [comment]
@enj ack, thanks!

**Comment by liggitt** [comment]
it looks like go added custom verification in https://tip.golang.org/pkg/crypto/tls/#example_Config_verifyConnection in 1.15, which is *almost* what we want, but sadly, this does not populate VerifiedChains:

```
	// Note that when certificates are not handled by the default verifier
	// ConnectionState.VerifiedChains will be nil.
```

**Comment by liggitt** [comment]
it's possible https://github.com/golang/go/issues/32406 might allow for request-level context to enable this (trying to follow the proposed implementation in https://go-review.googlesource.com/c/go/+/181097 to understand what capabilities it could give us)

**Comment by tallclair** [comment]
> This cert cache needs to support audit annotations, i.e. #90140.

@enj what did you have in mind? It doesn't look like the the x509 authenticator adds any annotations today. Would you just expect an annotation indicating that the cache was used to validate it?
